hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -0,0 +1,817 @@
+import org.rocksdb.CompactionJobInfo;
+import org.rocksdb.CompressionType;
+import org.rocksdb.DBOptions;
+import org.rocksdb.FlushOptions;
+import org.rocksdb.LiveFileMetaData;
+import org.rocksdb.Options;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.SstFileReader;
+import org.rocksdb.TableProperties;
+
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+// TODO
+//  1. Create a local instance of RocksDiff-local-RocksDB. This is the
+//  rocksDB that we can use for maintaining DAG and any other state. This is
+//  a per node state so it it doesn't have to go through RATIS anyway.
+//  2. Store fwd DAG in Diff-Local-RocksDB in Compaction Listener
+//  3. Store fwd DAG in Diff-Local-RocksDB in Compaction Listener
+//  4. Store last-Snapshot-counter/Compaction-generation-counter in Diff-Local
+//  -RocksDB in Compaction Listener
+//  5. System Restart handling. Read the DAG from Disk and load it in memory.
+//  6. Take the base snapshot. All the SST file nodes in the base snapshot
+//  should be arked with that Snapshot generation. Subsequently, all SST file
+//  node should have a snapshot-generation count and Compaction-generation
+//  count.
+//  6. DAG based SST file pruning. Start from the oldest snapshot and we can
+//  unlink any SST
+//  file from the SaveCompactedFilePath directory that is reachable in the
+//  reverse DAG.
+//  7. DAG pruning : For each snapshotted bucket, We can recycle the part of
+//  the DAG that is older than the predefined policy for the efficient snapdiff.
+//  E.g. we may decide not to support efficient snapdiff from any snapshot that
+//  is older than 2 weeks.
+//  Note on 8. & 9 .
+//  ==================
+//  A simple handling is to just iterate over all keys in keyspace when the
+//  compaction DAG is lost, instead of optimizing every case. And start
+//  Compaction-DAG afresh from the latest snapshot.
+//  --
+//  8. Handle bootstrapping rocksDB for a new OM follower node
+//      - new node will receive Active object store as well as all existing
+//      rocksDB checkpoints.
+//      - This bootstrapping should also receive the compaction-DAG information
+//  9. Handle rebuilding the DAG for a lagging follower. There are two cases
+//      - recieve RATIS transactions to replay. Nothing needs to be done in
+//      thise case.
+//      - Getting the DB sync. This case needs to handle getting the
+//      compaction-DAG information as well.
+//
+//
+/**
+ *  RocksDBCheckpointDiffer class.
+ */
+//CHECKSTYLE:OFF
+@SuppressWarnings({"NM_METHOD_NAMING_CONVENTION"})
+public class RocksDBCheckpointDiffer {
+  private final String rocksDbPath;
+  private String cpPath;
+  private final String CFDBPATH;
+  private final String SAVE_COMPACTED_FILE_PATH;
+  private final int MAX_SNAPSHOTS;
```

Comment:
These are not constants, we should have them in camel case.

Reply:
done